File: 125_ValidPalindrome.java
class Solution {
    boolean isAlpha(char c) {
        if ('A' <= c && c <= 'Z')
            return true;
        else if ('a' <= c && c <= 'z')
            return true;
        else if ('0' <= c && c <= '9')
            return true;
        return false;
    }

    public boolean isPalindrome(String s) {
        int i = 0;
        int j = s.length() - 1;

        while (i < j) {
            while (!isAlpha(s.charAt(i)) && i < j) i++;
            while (!isAlpha(s.charAt(j)) && i < j) j--;

            if(Character.toLowerCase(s.charAt(i)) != Character.toLowerCase(s.charAt(j))) {
                return false;
            }
            i++;
            j--;
        }

        return true;
    }
}

File: c++/125_ValidPalindrome.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        for (int i = 0, j = s.size() - 1; i < j; i++, j--) {
            while (i < j && !check(s[i])) i++;
            while (i < j && !check(s[j])) j--;
            if (i < j && tolower(s[i]) != tolower(s[j])) return false;
        }
        return true;
    }

    bool check(char c) {
        return c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z' || c >= '0' && c <= '9';
    }
};